qiskit/circuit/classicalfunction/boolean_expression.py

```diff
@@ -0,0 +1,122 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2021.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""A quantum oracle constructed from a logical expression or a string in the DIMACS format."""
+
+from os.path import basename, isfile
+
+from qiskit.exceptions import MissingOptionalLibraryError
+from .classical_element import ClassicalElement
+from .utils import HAS_TWEEDLEDUM
+
+
+class BooleanExpression(ClassicalElement):
+    """The Boolean Expression gate."""
+
+    def __init__(self, expression: str, name: str = None) -> None:
+        """
+        Args:
+            expression (str): The logical expression string.
+            name (str): Optional. Instruction gate name. Otherwise part of
+                        the expression is going to be used.
+        Raises:
+            MissingOptionalLibraryError: If tweedledum is not installed. Tweedledum is required.
+        """
+        if not HAS_TWEEDLEDUM:
+            raise MissingOptionalLibraryError(
+                libname='tweedledum',
+                name='BooleanExpression compiler',
+                pip_install='pip install tweedledum')
+        from tweedledum import BoolFunction
+        self._tweedledum_bool_expression = BoolFunction.from_expression(expression)
+        self.qregs = None  # TODO: Probably from self._tweedledum_bool_expression._signature
+
+        short_expr_for_name = (expression[:10] + '...') if len(expression) > 13 else expression
+        num_qubits = (self._tweedledum_bool_expression.num_outputs() +
+                      self._tweedledum_bool_expression.num_inputs())
+        super().__init__(name or short_expr_for_name, num_qubits=num_qubits, params=[])
+
+    def simulate(self, bitstring: str) -> bool:
+        """Evaluate the expression on a bitstring.
+
+        This evaluation is done classically.
+
+        Args:
+            bitstring: The bitstring for which to evaluate.
+
+        Returns:
+            bool: result of the evaluation.
+        """
+        from tweedledum import BitVec
+        bits = []
+        for bit in bitstring:
+            bits.append(BitVec(1, bit))
+        return bool(self._tweedledum_bool_expression.simulate(*bits))
+
+    def synth(self, registerless=True, synthesizer=None):
```

Comment:
Could you move the typehints to the signature?

Reply:
done in c025c487d